hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileReader.java

```diff
@@ -0,0 +1,178 @@
+      return Optional.empty();
+    }
+    BlockIndexEntry blockToRead = blockIndexEntryList.get(rootLevelBlockIndex);
+    HFileBlockReader blockReader = new HFileBlockReader(
+        context, stream, blockToRead.getOffset(), blockToRead.getOffset() + (long) blockToRead.getSize());
+    HFileDataBlock dataBlock = (HFileDataBlock) blockReader.nextBlock(HFileBlockType.DATA);
+    return seekToKeyInBlock(dataBlock, key);
+  }
+
+  /**
+   * Reads the HFile major version from the input.
+   *
+   * @param bytes  Input data.
+   * @param offset Offset to start reading.
+   * @return Major version of the file.
+   */
+  public static int readMajorVersion(byte[] bytes, int offset) {
+    int ch1 = bytes[offset] & 0xFF;
+    int ch2 = bytes[offset + 1] & 0xFF;
+    int ch3 = bytes[offset + 2] & 0xFF;
+    return ((ch1 << 16) + (ch2 << 8) + ch3);
+  }
+
+  /**
+   * Reads and parses the HFile trailer.
+   *
+   * @param stream   HFile input.
+   * @param fileSize HFile size.
+   * @return {@link HFileTrailer} instance.
+   * @throws IOException upon error.
+   */
+  private static HFileTrailer readTrailer(FSDataInputStream stream,
+                                          long fileSize) throws IOException {
+    int bufferSize = HFileTrailer.getTrailerSize();
+    long seekPos = fileSize - bufferSize;
+    if (seekPos < 0) {
+      // It is hard to imagine such a small HFile.
+      seekPos = 0;
+      bufferSize = (int) fileSize;
+    }
+    stream.seek(seekPos);
+
+    byte[] byteBuff = new byte[bufferSize];
+    stream.readFully(byteBuff);
+
+    int majorVersion = readMajorVersion(byteBuff, bufferSize - 3);
+    int minorVersion = byteBuff[bufferSize - 3];
+
+    HFileTrailer trailer = new HFileTrailer(majorVersion, minorVersion);
+    trailer.deserialize(new DataInputStream(new ByteArrayInputStream(byteBuff)));
+    return trailer;
+  }
+
+  /**
+   * Searches the block that may contain the lookup key based the starting keys
+   * of all blocks (sorted in the input list), using binary search.
+   *
+   * @param lookUpKey The key to lookup.
+   * @return Block index in the input. An index outside the range of input means the key does not
+   * exist in the HFile.
+   */
+  private int searchBlockByKey(BlockIndexEntry lookUpKey) {
+    int pos = Collections.binarySearch(blockIndexEntryList, lookUpKey);
+    // pos is between -(blockKeys.length + 1) to blockKeys.length - 1, see
+    // binarySearch's javadoc.
+
+    if (pos >= 0) {
+      // This means this is an exact match with an element of blockKeys.
+      assert pos < blockIndexEntryList.size();
+      return pos;
+    }
+
+    // Otherwise, pos = -(i + 1), where blockKeys[i - 1] < key < blockKeys[i],
+    // and i is in [0, blockKeys.length]. We are returning j = i - 1 such that
+    // blockKeys[j] <= key < blockKeys[j + 1]. In particular, j = -1 if
+    // key < blockKeys[0], meaning the file does not contain the given key.
+
+    int i = -pos - 1;
```

Comment:
this kind of stuff - seems correct. but when we get down to debugging this day-to-day, probably a higher level data structure helps?

Reply:
Fixed now by using TreeMap.